# System for access segmentation

##### imports

In [1]:
import numpy as np
import pandas as pd
import os
from sklearn.preprocessing import StandardScaler

#### Get data from results/data folder

In [2]:
data = pd.data = pd.read_excel("../server/public/data/User-624921c5e1c4d602ddd19695-1649206670127.xlsx", skiprows=1)

In [3]:
data

user_id                     position  \
0       AALISMNA          POSTDOCTORAL FELLOW   
1       AALISMNA          POSTDOCTORAL FELLOW   
2       AALISMNA          POSTDOCTORAL FELLOW   
3       AALISMNA          POSTDOCTORAL FELLOW   
4       AALISMNA          POSTDOCTORAL FELLOW   
...          ...                          ...   
199530   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199531   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199532   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199533   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199534   ZYBINSA  SENIOR TECHNICAL SPECIALIST   

                                  org_structure  \
0       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
1       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
2       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
3       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
4       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
...                                         ...   
199530                               INORGANICS   
199531                               INORGANICS   
199532                               INORGANICS   
199533                               INORGANICS   
199534                               INORGANICS   

                                             access_name  
0                               PRIV:ROLE:GOOGLE:ACCOUNT  
1                              PRIV:ROLE:BAMP900:PB:USER  
2                          PRIV:ROLE:BAMP900:RMS:HR_USER  
3                         PRIV:ROLE:BGWP960:PB:FIORI_FLP  
4                       PRIV:ROLE:BAMP900:FI:TV:TRAVELER  
...                                                  ...  
199530     PRIV:ROLE:BAMP900:HR:ESS:CLAIMS_REIMBURSEMENT  
199531     PRIV:ROLE:BGWP960:IT:GW:PASSWORD_SELF_SERVICE  
199532  PRIV:ROLE:BAMP900:IT:BUSINESS_JOB_RELEASE_ACCESS  
199533  PRIV:ROLE:BGWP960:KS:KITS_CONFERENC_ROOM_BOOKING  
199534  PRIV:ROLE:BGWP960:KS:KITS_EMPLOYEE_EQUIPMENT_GRP  

[199535 rows x 4 columns]

In [4]:
data.columns = ["user_id", "job", "department", "privileges_data"]

In [5]:
data = data.dropna()

In [6]:
data

user_id                          job  \
0       AALISMNA          POSTDOCTORAL FELLOW   
1       AALISMNA          POSTDOCTORAL FELLOW   
2       AALISMNA          POSTDOCTORAL FELLOW   
3       AALISMNA          POSTDOCTORAL FELLOW   
4       AALISMNA          POSTDOCTORAL FELLOW   
...          ...                          ...   
199530   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199531   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199532   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199533   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199534   ZYBINSA  SENIOR TECHNICAL SPECIALIST   

                                     department  \
0       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
1       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
2       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
3       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
4       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG   
...                                         ...   
199530                               INORGANICS   
199531                               INORGANICS   
199532                               INORGANICS   
199533                               INORGANICS   
199534                               INORGANICS   

                                         privileges_data  
0                               PRIV:ROLE:GOOGLE:ACCOUNT  
1                              PRIV:ROLE:BAMP900:PB:USER  
2                          PRIV:ROLE:BAMP900:RMS:HR_USER  
3                         PRIV:ROLE:BGWP960:PB:FIORI_FLP  
4                       PRIV:ROLE:BAMP900:FI:TV:TRAVELER  
...                                                  ...  
199530     PRIV:ROLE:BAMP900:HR:ESS:CLAIMS_REIMBURSEMENT  
199531     PRIV:ROLE:BGWP960:IT:GW:PASSWORD_SELF_SERVICE  
199532  PRIV:ROLE:BAMP900:IT:BUSINESS_JOB_RELEASE_ACCESS  
199533  PRIV:ROLE:BGWP960:KS:KITS_CONFERENC_ROOM_BOOKING  
199534  PRIV:ROLE:BGWP960:KS:KITS_EMPLOYEE_EQUIPMENT_GRP  

[199511 rows x 4 columns]

In [7]:
user_data = data.drop("privileges_data", axis=1)

In [8]:
user_data

user_id                          job  \
0       AALISMNA          POSTDOCTORAL FELLOW   
1       AALISMNA          POSTDOCTORAL FELLOW   
2       AALISMNA          POSTDOCTORAL FELLOW   
3       AALISMNA          POSTDOCTORAL FELLOW   
4       AALISMNA          POSTDOCTORAL FELLOW   
...          ...                          ...   
199530   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199531   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199532   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199533   ZYBINSA  SENIOR TECHNICAL SPECIALIST   
199534   ZYBINSA  SENIOR TECHNICAL SPECIALIST   

                                     department  
0       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG  
1       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG  
2       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG  
3       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG  
4       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG  
...                                         ...  
199530                               INORGANICS  
199531                               INORGANICS  
199532                               INORGANICS  
199533                               INORGANICS  
199534                               INORGANICS  

[199511 rows x 3 columns]

In [9]:
privileges_data = pd.DataFrame(data["privileges_data"])

In [10]:
privileges_data.privileges_data = privileges_data["privileges_data"].str[10:]

I saw anomaly in data where in some of the records there is ::(double) symbols so i need to remove it so the data is clear

In [11]:
for indx in privileges_data[privileges_data["privileges_data"].str.startswith(":")].index.tolist():
    privileges_data.loc[indx].privileges_data = privileges_data.loc[indx].privileges_data[1:]

In [12]:
len(privileges_data.loc[privileges_data["privileges_data"].str.startswith(":")].privileges_data)

0

In [13]:
privileges_data

privileges_data
0                               GOOGLE:ACCOUNT
1                              BAMP900:PB:USER
2                          BAMP900:RMS:HR_USER
3                         BGWP960:PB:FIORI_FLP
4                       BAMP900:FI:TV:TRAVELER
...                                        ...
199530     BAMP900:HR:ESS:CLAIMS_REIMBURSEMENT
199531     BGWP960:IT:GW:PASSWORD_SELF_SERVICE
199532  BAMP900:IT:BUSINESS_JOB_RELEASE_ACCESS
199533  BGWP960:KS:KITS_CONFERENC_ROOM_BOOKING
199534  BGWP960:KS:KITS_EMPLOYEE_EQUIPMENT_GRP

[199511 rows x 1 columns]

In [14]:
privileges_data['system_name'], privileges_data['priv'] = privileges_data['privileges_data'].str.split(':', 1, expand=True)[0], privileges_data['privileges_data'].str.split(':', 1, expand=True)[1]

In [15]:
privileges_data = privileges_data.drop('privileges_data', axis=1)

In [16]:
privileges_data

system_name                            priv
0           GOOGLE                         ACCOUNT
1          BAMP900                         PB:USER
2          BAMP900                     RMS:HR_USER
3          BGWP960                    PB:FIORI_FLP
4          BAMP900                  FI:TV:TRAVELER
...            ...                             ...
199530     BAMP900     HR:ESS:CLAIMS_REIMBURSEMENT
199531     BGWP960     IT:GW:PASSWORD_SELF_SERVICE
199532     BAMP900  IT:BUSINESS_JOB_RELEASE_ACCESS
199533     BGWP960  KS:KITS_CONFERENC_ROOM_BOOKING
199534     BGWP960  KS:KITS_EMPLOYEE_EQUIPMENT_GRP

[199511 rows x 2 columns]

In [17]:
user_data['system_name'] = privileges_data['system_name']
user_data['priv'] = privileges_data['priv']

In [18]:
job="POSTDOCTORAL FELLOW"
department="BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG"

segmentated_data = user_data.where((user_data.job == job) & (user_data.department == department))
segmentated_data.dropna(inplace=True)

In [19]:
segmentated_data

user_id                  job  \
0       AALISMNA  POSTDOCTORAL FELLOW   
1       AALISMNA  POSTDOCTORAL FELLOW   
2       AALISMNA  POSTDOCTORAL FELLOW   
3       AALISMNA  POSTDOCTORAL FELLOW   
4       AALISMNA  POSTDOCTORAL FELLOW   
...          ...                  ...   
198873      ZIAA  POSTDOCTORAL FELLOW   
198874      ZIAA  POSTDOCTORAL FELLOW   
198875      ZIAA  POSTDOCTORAL FELLOW   
198876      ZIAA  POSTDOCTORAL FELLOW   
198877      ZIAA  POSTDOCTORAL FELLOW   

                                     department system_name  \
0       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG      GOOGLE   
1       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BAMP900   
2       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BAMP900   
3       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BGWP960   
4       BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BAMP900   
...                                         ...         ...   
198873  BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BGWP960   
198874  BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BGWP960   
198875  BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BAMP900   
198876  BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BGWP960   
198877  BIOLOGICAL, ENVIRONMENTAL SCIENCE & ENG     BGWP960   

                                  priv  
0                              ACCOUNT  
1                              PB:USER  
2                          RMS:HR_USER  
3                         PB:FIORI_FLP  
4                       FI:TV:TRAVELER  
...                                ...  
198873     HR:GW:CLAIMS_TRAVEL_OLD_TIL  
198874     IT:GW:PASSWORD_SELF_SERVICE  
198875  IT:BUSINESS_JOB_RELEASE_ACCESS  
198876  KS:KITS_CONFERENC_ROOM_BOOKING  
198877  KS:KITS_EMPLOYEE_EQUIPMENT_GRP  

[2199 rows x 5 columns]

In [20]:
avarege_count = int(segmentated_data.priv.value_counts().sum() / len(segmentated_data.priv.value_counts()))
count = 0
for p_c in segmentated_data.priv.value_counts():
    if p_c < avarege_count:
        break
    count += 1
priv_list = segmentated_data.priv.value_counts().index.tolist()[:count]

In [21]:
priv_list

['KAUST:END_USERS',
 'IT:UWL:ENDUSER',
 'GRAC:BUSINESS_ENDUSERS',
 'GRAC:END_USERS_BASE',
 'RMS:HR_USER',
 'SM:ITSM:SELF_SERVICE',
 'IT:GW:PASSWORD_SELF_SERVICE',
 'IT:FIORI_INBOX_APPROVAL',
 'KS:KITS_EMPLOYEE_EQUIPMENT_GRP',
 'IT:GW:END_USER',
 'SM:ITSM:CREATE_INTERACTION',
 'GA:GASC_MYREQUEST',
 'HR:ESS:CLAIMS_REIMBURSEMENT',
 'IT:GW:KAUST_ACCES_REQUEST',
 'KS:KITS_CONFERENC_ROOM_BOOKING',
 'KS:KITS_SECURITY_SERVICES',
 'GA:GASC_REQUESTER',
 'HR:GW:EMP_SELF_SERVICE',
 'IT:BUSINESS_JOB_RELEASE_ACCESS',
 'PB:FIORI_FLP',
 'ACCOUNT',
 'FI:TV:TRAVELER',
 'KS:KITS_EMAIL_SERVICES',
 'PB:USER',
 'HR:ESS:EMP_SELF_SERVICE',
 'RMS:TV:HR_USER',
 'SLCM:RM_END_USER',
 'IT:GW:USER_NOTIF',
 'IT:GW:USER_REPORTS',
 'IT:GW:USER_WORKFLOW']

### Get information about dataset properties

#### Get columns

In [22]:
user_data.columns

Index(['user_id', 'job', 'department', 'system_name', 'priv'], dtype='object')

#### Get unique attributes in every column

In [23]:
for c in user_data.columns:
    print(c, len(user_data[c].value_counts()))

user_id 6784
job 1581
department 463
system_name 9
priv 2049


#### Get unique privileges count for every system 

In [24]:
for s in user_data.system_name.unique():
    print(s, len(user_data[user_data.system_name == s].priv.unique()))

GOOGLE 1
BAMP900 1751
BGWP960 71
SMP600 22
BGRP940 7
EMP980 57
BMP950 210
BCMP970 2
IT 1


#### get how many times system occurs in data

In [25]:
user_data.system_name.value_counts()

BGWP960    79886
BAMP900    77340
SMP600     16348
BGRP940    14272
EMP980      5542
BMP950      3642
GOOGLE      2406
BCMP970       66
IT             9
Name: system_name, dtype: int64

We can conclude that BCMP970 and IT systems contains special privileges for just one certain department and job

#### get data about privileges. Are there privileges which are the same for various systems

In [26]:
len(user_data.priv.unique())

2049

### Prepare data for clustering

In simple words we don't need data for job and department to cluster data, because clustering is performed on privileges for all users.

So we will create new table with user id - user_id, and all of the privileges as attributes with values of 0 or 1. 0 - don't have that role, 1 - have that role.

In [27]:
user_privs_data = pd.DataFrame()

In [28]:
user_privs_data['user_id'] = user_data.user_id.unique()

In [29]:
user_privs_data

user_id
0     AALISMNA
1       AAWARE
2      ABABTAM
3      ABALAMP
4      ABASISN
...        ...
6779  ZULFIQMH
6780    ZUPINF
6781  ZURAYQOA
6782   ZWAWIFM
6783   ZYBINSA

[6784 rows x 1 columns]

Create all privileges(categorical data) compatabel for clustering by transforming all privs into numerical data.

In [30]:
#priv_dummies = pd.DataFrame()

In [31]:
#priv_dummies['user_id'] = user_data.user_id

In [32]:
#priv_dummies = pd.concat([priv_dummies,pd.get_dummies(user_data.priv)], axis=1)

In [33]:
#priv_dummies.groupby('user_id').sum()

After that a problem occurs. The shape of dummies is (199511 × 2049) but our data for clustering - user_privs_data is with shape (6784 x 1)

In [34]:
user_privs_data.reset_index()

NameError: name 'priv' is not defined